qiskit_nature/circuit/library/initial_states/fermionic_gaussian_states.py

```diff
@@ -0,0 +1,361 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Fermionic Gaussian states."""
+
+from typing import Iterable, List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from qiskit import QuantumRegister, QuantumCircuit
+from qiskit.circuit import Gate, Qubit
+from qiskit.circuit.library import RZGate, XGate, XYGate
+
+from qiskit_nature.converters.second_quantization import QubitConverter
+from qiskit_nature.mappers.second_quantization import JordanWignerMapper
+
+
+class SlaterDeterminant(QuantumCircuit):
```

Comment:
I would split this out into its own file

Reply:
Splitting the two classes into separate files would result in a lot of code duplication because many of the private helper functions would need to be copied to both files. What would be the benefit?

Reply:
So far we have always been splitting each class into its own file. I only noticed the common module-level functions later in my review. We don't have a lot of module level functions in Qiskit Nature and if we do we sometimes create a `utils` module. See for example here: https://github.com/Qiskit/qiskit-nature/tree/main/qiskit_nature/circuit/library/ansatzes/utils

@woodsp-ibm @pbark what do you think?

Reply:
It might make sense to factor out and expose some of the more generally useful private functions as utilities, but I don't think that's appropriate for the highly specific helper functions like `_prepare_slater_determinant_jordan_wigner`. Is there a benefit to enforcing that every class go in its own file?

Reply:
For me the two circuits would seem reasonable separate modules if only that when browsing the code base the fact that there are these 2 different circuits that can be used stands out from just looking at the files and their names (hopefully!). The utility methods can be with the class that needs them - for the common ones, which seem to be matrix ones  they could simply be put in another file. I am not sure it would need a whole new utils folder here, and so far Nature does not have a root level utils where perhaps more general matrix manipulation functions might have been situated. Perhaps arranging the structure this way then if another similar circuit was created in the future it would facilitate that better and the whoever supplied it would not be tempted to just add it to the same file.

Reply:
> For me the two circuits would seem reasonable separate modules if only that when browsing the code base the fact that there are these 2 different circuits that can be used stands out from just looking at the files and their names (hopefully!).

It's much more reliable in general to discover different classes by browsing the parsed documentation rather than making assumptions about how code is distributed between source files.

> The utility methods can be with the class that needs them - for the common ones, which seem to be matrix ones they could simply be put in another file.

I agree. The (uncommon) utility methods here belong with the class that needs them. Since in this case they are needed by both classes, both classes belong in the same file.

Reply:
Very well. I guess I can put the helper functions in a private file without exposing that file to the public API.

Reply:
Done.